### ISSUE TITLE
Netdata_cgroup_cpu_limit is absent if limits.cpu=1 in the LXC container configuration

### ISSUE THREAD

**Issue by vpr-ossteam** [netdata/netdata#10670]
##### Bug report summary

When limits.cpu is upper than 1 in the LXC container configuration netdata_cgroup_cpu_limit metric is present, but if limits.cpu=1 the metric is absent.  

##### OS / Environment

```
~$  uname -a; grep -Hv "^#" /etc/*release
Linux access 4.15.0-118-generic #119-Ubuntu SMP Tue Sep 8 12:30:01 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
/etc/lsb-release:DISTRIB_ID=Ubuntu
/etc/lsb-release:DISTRIB_RELEASE=18.04
/etc/lsb-release:DISTRIB_CODENAME=bionic
/etc/lsb-release:DISTRIB_DESCRIPTION="Ubuntu 18.04.5 LTS"
/etc/os-release:NAME="Ubuntu"
/etc/os-release:VERSION="18.04.5 LTS (Bionic Beaver)"
/etc/os-release:ID=ubuntu
/etc/os-release:ID_LIKE=debian
/etc/os-release:PRETTY_NAME="Ubuntu 18.04.5 LTS"
/etc/os-release:VERSION_ID="18.04"
/etc/os-release:HOME_URL="https://www.ubuntu.com/"
/etc/os-release:SUPPORT_URL="https://help.ubuntu.com/"
/etc/os-release:BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
/etc/os-release:PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
/etc/os-release:VERSION_CODENAME=bionic
/etc/os-release:UBUNTU_CODENAME=bionic
```

##### Netdata version

```
~# netdata -V
netdata v1.29.0
```

##### Steps To Reproduce

1. Set limits.cpu=2 or upper in the LXC container configuration
2. Check that the netdata_cgroup_cpu_limit metric is present
3. Set limits.cpu=1 and check that netdata_cgroup_cpu_limit metric is absent

##### Expected behavior

The value of limits.cpu mustn't change the behavior of Netdata and netdata_cgroup_cpu_limit metric must be present by anyway.

**Comment by ilyam8** [comment]
Hi @vpr-ossteam 



(likely) Related part of the code 

https://github.com/netdata/netdata/blob/9c768c691100998bc7910aefdd9ae4ff79417763/collectors/cgroups.plugin/sys_fs_cgroup.c#L3137-L3222

@vlvkobal help me to debug it.

There are some conditions to get that metric collected. What are they?

**Comment by ilyam8** [comment]
Ok, i see we use

- `/sys/fs/cgroup/cpuacct/**/cpu.cfs_quota_us` which is the total available run-time within a period (in microseconds)
- `/sys/fs/cgroup/cpuacct/**/cpu.cfs_period_us` which is the length of a period (in microseconds)
- `/sys/fs/cgroup/cpuacct/**/cpuset.cpus` which is a list of CPU cores to which a cgroup has access.

I did 

```cmd
docker run -d --rm --name nginx_no_cpu_limits nginx
docker run -d --rm --cpus=1 --name nginx_cpu_limits nginx
```

I see

```
# http://127.0.0.1:19999/api/v1/allmetrics?format=prometheus&source=raw&variables=true
netdata_cgroup_nginx_no_cpu_limits_cpu_cpu_limit 600.0000000 1615310665000
netdata_cgroup_nginx_cpu_limits_cpu_cpu_limit 100.0000000 1615310665000
```

**Comment by vpr-ossteam** [comment]
@ilyam8, I have no cpuset.cpus on my system (as you wrote, this is ok). In cpu.cfs_quota_us and cpu.cfs_period_us I've seen this:

```
~# cat /sys/fs/cgroup/cpuacct/**/cpu.cfs_quota_us
-1
-1
-1
~# cat /sys/fs/cgroup/cpuacct/**/cpu.cfs_period_us
100000
100000
100000
```
All LXC containers on this server have limits.cpu: "1" parameter in their configs.

**Comment by ilyam8** [comment]
Well, you have no cpu time restrictions. 

> cpu.cfs_quota_us

>  Setting the value in cpu.cfs_quota_us to -1 indicates that the cgroup does not adhere to any CPU time restrictions. This is also the default value for every cgroup (except the root cgroup).


I googled and found this article [setting-cpu-resource-limits-with-lxc](https://www.jamescoyle.net/how-to/2532-setting-cpu-resource-limits-with-lxc).

Trying to understand how lxc cpu limits commands translates into cgroups limits

CPU management is done in 1 of 4 ways:
 1. Number of CPUs
 2. Specific cores
 3. Capped share
 4. Limited time share


`limits.cpu` is 1 and 2 (could be an integer value or comma separated list with dashes)
`limit.cpu.allowance` is 3 and 4 (could be a percentage or allowed time/period)


@vpr-ossteam could you check same values for any LXC container with limits.cpu > 1 ?

**Comment by ilyam8** [comment]
I expected that `limits.cpu=N` uses `cpu.cfs_quota_us` and `cpu.cfs_period_us` to set the number of CPU cores that LXC can use with this container.

[Same as docker](https://docs.docker.com/config/containers/resource_constraints/#configure-the-default-cfs-scheduler)

> --cpus=<value>

> Specify how much of the available CPU resources a container can use. 

> For instance, if the host machine has two CPUs and you set --cpus="1.5", the container is guaranteed at most one and a half of the CPUs. This is the equivalent of setting --cpu-period="100000" and --cpu-quota="150000".

**Comment by n0guest** [comment]
> CPU management is done in 1 of 4 ways:
> 
> 1. Number of CPUs
> 2. Specific cores
> 3. Capped share
> 4. Limited time share

Yep, these are quite a mess. Because we have three cgroups-v1 controllers (cpu, cpuacct and cpuset). On top of this we have not-so-clear [LXD internal logic](https://github.com/lxc/lxd/blob/508fd2c608c617b59a11178178fb5b5b4038774f/lxd/instance/drivers/driver_lxc.go#L1153-L1176) which translates their own options into cgroups-v1 parameters. But the initial questions, imo, remains the same. If we have `limits.cpu=1`, why Netdata doesn't show it as `netdata_cgroup_cpu_limit=1` (in export for Prometheus)?

Here is an example. We have a container with following limits in LXD:

```
# lxc config show demo | grep limits.cpu
  limits.cpu: "1"
  limits.cpu.allowance: 90ms/100ms
```

Which translates into this:

```
# cat /sys/fs/cgroup/cpu/lxc/demo/cpu.cfs_period_us 
100000
# cat /sys/fs/cgroup/cpu/lxc/demo/cpu.cfs_quota_us 
90000
# cat /sys/fs/cgroup/cpuset/lxc/demo/cpuset.cpus
0
# cat /sys/fs/cgroup/cpu/lxc/demo/cpu.shares 
1024
```

So, target control cgroup (i.e. container) can see only one CPU core and use up to 90% if it's power. In other words right now we have our limits, but Netdata still doesn't show required chart (same goes for mentioned metrics). The only way to force Netdata to export discussed metric is to make more than one CPU core visible:

```
# lxc config show demo | grep limits.cpu
  limits.cpu: "2"
  limits.cpu.allowance: 90ms/100ms

# cat /sys/fs/cgroup/cpuset/lxc/demo/cpuset.cpus
0-1
# cat /sys/fs/cgroup/cpu/lxc/demo/cpu.cfs_quota_us 
90000
# cat /sys/fs/cgroup/cpu/lxc/demo/cpu.cfs_period_us 
100000
# cat /sys/fs/cgroup/cpu/lxc/demo/cpu.shares 
1024
```

@ilyam8, probably this behavior is coded somewhere [here](https://github.com/netdata/netdata/blob/9c768c691100998bc7910aefdd9ae4ff79417763/collectors/cgroups.plugin/sys_fs_cgroup.c#L3161-L3173)?

PS: Is there anything else is needed? I.e. additional checks, etc?

**Comment by ilyam8** [comment]
Yeah, @n0guest, you are right. 

```cmd
docker run -d --rm --cpuset-cpus=0 --name nginx nginx
```

No cpu limits.

**Comment by ilyam8** [comment]
I guess we always one core behind. That is why there is no limit if it is only one core 😅  

https://github.com/netdata/netdata/blob/9c768c691100998bc7910aefdd9ae4ff79417763/collectors/cgroups.plugin/sys_fs_cgroup.c#L2937-L2950

**Comment by ilyam8** [comment]
@vpr-ossteam @n0guest thx for reporting/helping!

(i hope) Fixed in #10757

**Comment by n0guest** [comment]
> @vpr-ossteam @n0guest thx for reporting/helping!
> 
> (i hope) Fixed in #10757

Thank you. I hope it will be included in the upcoming 1.29.4 release.
